portal-api/src/v1/users/sign-in-link.service.js

```diff
@@ -0,0 +1,69 @@
+const { ObjectId } = require('mongodb');
+const crypto = require('node:crypto');
+const sendEmail = require('../email');
+const { EMAIL_TEMPLATE_IDS, SIGN_IN_LINK_EXPIRY_MINUTES } = require('../../constants');
+const db = require('../../drivers/db-client');
+
+module.exports.createAndEmailSignInLink = async (user) => {
+  const {
+    _id: userId,
+    email: userEmail,
+    firstname: userFirstName,
+    surname: userLastName,
+  } = user;
+
+  const signInCode = await createSignInCode();
+  await saveSignInCodeHashAndSalt({ userId, signInCode });
+  return sendSignInLinkEmail({
+    userEmail,
+    userFirstName,
+    userLastName,
+    // TODO DTFS2-6750: introduce an environment variable to replace localhost
+    signInLink: `http://localhost/login/sign-in-link?t=${signInCode}`,
+  });
+};
+
+async function createSignInCode() {
+  try {
+    return crypto.randomBytes(32).toString('hex');
+  } catch (e) {
+    const error = new Error('Failed to create a sign in code.');
+    error.cause = e;
+    throw error;
+  }
+}
+
+async function saveSignInCodeHashAndSalt({ userId, signInCode }) {
+  try {
+    const salt = crypto.randomBytes(64);
+    const hash = crypto.pbkdf2Sync(signInCode, salt, 210000, 64, 'sha512').toString('hex');
+
+    // TODO DTFS2-6750: should we introduce a repository here?
```

Comment:
I'd love a repository here -- Is the debt too high to be covered as part of this ticket (ideally, I'm guessing we'd have a users repo)?

Reply:
Great to know that you think it's worth doing too. The idea I was struggling with was how far I go on making a repository that works just for the current functionality v.s. a repository that works for all user updates (and using that in any existing code that updates a user)

I'll have a look at this and see what I can come up with